odoo/http.py

```diff
@@ -1,836 +1,1086 @@
-           to arbitrary responses. Anything returned (except None) will
-           be used as response."""
-        try:
-            return super(JsonRequest, self)._handle_exception(exception)
-        except Exception:
-            if not isinstance(exception, SessionExpiredException):
-                if exception.args and exception.args[0] == "bus.Bus not available in test mode":
-                    _logger.info(exception)
-                elif isinstance(exception, (odoo.exceptions.UserError,
-                                            werkzeug.exceptions.NotFound)):
-                    _logger.warning(exception)
-                else:
-                    _logger.exception("Exception during JSON request handling.")
-            error = {
-                'code': 200,
-                'message': "Odoo Server Error",
-                'data': serialize_exception(exception),
-            }
-            if isinstance(exception, werkzeug.exceptions.NotFound):
-                error['http_status'] = 404
-                error['code'] = 404
-                error['message'] = "404: Not Found"
-            if isinstance(exception, AuthenticationError):
-                error['code'] = 100
-                error['message'] = "Odoo Session Invalid"
-            if isinstance(exception, SessionExpiredException):
-                error['code'] = 100
-                error['message'] = "Odoo Session Expired"
-            return self._json_response(error=error)
-
-    def dispatch(self):
-        rpc_request_flag = rpc_request.isEnabledFor(logging.DEBUG)
-        rpc_response_flag = rpc_response.isEnabledFor(logging.DEBUG)
-        if rpc_request_flag or rpc_response_flag:
-            endpoint = self.endpoint.method.__name__
-            model = self.params.get('model')
-            method = self.params.get('method')
-            args = self.params.get('args', [])
-
-            start_time = time.time()
-            start_memory = 0
-            if psutil:
-                start_memory = memory_info(psutil.Process(os.getpid()))
-            if rpc_request and rpc_response_flag:
-                rpc_request.debug('%s: %s %s, %s',
-                    endpoint, model, method, pprint.pformat(args))
-
-        result = self._call_function(**self.params)
-
-        if rpc_request_flag or rpc_response_flag:
-            end_time = time.time()
-            end_memory = 0
-            if psutil:
-                end_memory = memory_info(psutil.Process(os.getpid()))
-            logline = '%s: %s %s: time:%.3fs mem: %sk -> %sk (diff: %sk)' % (
-                endpoint, model, method, end_time - start_time, start_memory / 1024, end_memory / 1024, (end_memory - start_memory)/1024)
-            if rpc_response_flag:
-                rpc_response.debug('%s, %s', logline, pprint.pformat(result))
+    def _get_profiler_context_manager(self):
+        """ Return a context manager that combines a profiler and ``request``. """
```

Comment:
Docstring is out of date.

Reply:
Oh, and by the way, :+1+ for using `contextlib.nullcontext()`.  The former solution was mediocre IMHO.